crypto/crypto/src/noise.rs

```diff
@@ -487,54 +653,67 @@ impl NoiseSession {
             return Err(NoiseError::PayloadTooLarge);
         }
 
-        // encrypt
+        // encrypt in place
         let aead = Aes256Gcm::new(GenericArray::clone_from_slice(&self.write_key));
-
-        let msg_and_ad = Payload {
-            msg: plaintext,
-            aad: &[],
-        };
         let mut nonce = [0u8; 4].to_vec();
         nonce.extend_from_slice(&self.write_nonce.to_be_bytes());
         let nonce = GenericArray::from_slice(&nonce);
-        let ciphertext = aead.encrypt(nonce, msg_and_ad).unwrap(); // this API cannot fail
+
+        let authentication_tag = aead
+            .encrypt_in_place_detached(nonce, b"", plaintext)
+            .map_err(|_| NoiseError::Encrypt)?;
 
         // increment nonce
         self.write_nonce += 1;
 
-        //
-        Ok(ciphertext)
+        // return a subslice without the authentication tag
+        Ok(authentication_tag.to_vec())
     }
 
     /// decrypts a message from the other peer (post-handshake)
-    pub fn read_message(&mut self, ciphertext: &[u8]) -> Result<Vec<u8>, NoiseError> {
+    /// the function decrypts in place, and returns a subslice without the auth tag
+    pub fn read_message_in_place<'a>(
+        &mut self,
+        ciphertext: &'a mut [u8],
+    ) -> Result<&'a [u8], NoiseError> {
```

Comment:
if you think an API that takes ciphertext + tag is better, it would simplify the code on this side. But it could mean that we would just move the complexity on the caller side.

Reply:
it's probably fine for now like this.